src/encoder.rs

```diff
@@ -2303,44 +2309,63 @@ fn encode_partition_bottomup(seq: &Sequence, fi: &FrameInvariants, fs: &mut Fram
+                continue;
+            }
+            assert!(bsw == bsh);
+            if must_split {
+                let cbw = (fi.w_in_b - bo.x).min(bsw); // clipped block width, i.e. having effective pixels
+                let cbh = (fi.h_in_b - bo.y).min(bsh);
+                let mut split_vert = false;
+                let mut split_horz = false;
+                if cbw == bsw/2 && cbh == bsh { split_vert = true; }
+                if cbh == bsh/2 && cbw == bsw { split_horz = true; }
+                if !split_horz {
+                    if partition == PartitionType::PARTITION_HORZ { continue; };
+                }
+                if !split_vert {
+                    if partition == PartitionType::PARTITION_VERT { continue; };
+                }
+            }
+            cw.rollback(&cw_checkpoint);
+            w_pre_cdef.rollback(&w_pre_checkpoint);
+            w_post_cdef.rollback(&w_post_checkpoint);
 
-        rd_cost = 0.0;
+            subsize = bsize.subsize(partition);
+            let hbsw = subsize.width_mi(); // Half the block size width in blocks
+            let hbsh = subsize.height_mi(); // Half the block size height in blocks
+            let mut child_modes: Vec<RDOPartitionOutput> = Vec::new();
+            rd_cost = 0.0;
 
-        if bsize.gte(BlockSize::BLOCK_8X8) {
-            let w: &mut dyn Writer = if cw.bc.cdef_coded {w_post_cdef} else {w_pre_cdef};
-            let tell = w.tell_frac();
-            cw.write_partition(w, bo, partition, bsize);
-            rd_cost = (w.tell_frac() - tell) as f64 * get_lambda(fi, seq.bit_depth)/ ((1 << OD_BITRES) as f64);
-        }
+            if bsize.gte(BlockSize::BLOCK_8X8) {
+                let w: &mut dyn Writer = if cw.bc.cdef_coded {w_post_cdef} else {w_pre_cdef};
+                let tell = w.tell_frac();
+                cw.write_partition(w, bo, partition, bsize);
+                rd_cost = (w.tell_frac() - tell) as f64 * get_lambda(fi, seq.bit_depth)/ ((1 << OD_BITRES) as f64);
+            }
 
-        let four_partitions = [
-            bo,
-            &BlockOffset{ x: bo.x + hbsw as usize, y: bo.y },
-            &BlockOffset{ x: bo.x, y: bo.y + hbsh as usize },
-            &BlockOffset{ x: bo.x + hbsw as usize, y: bo.y + hbsh as usize }
-        ];
-        let partitions = get_sub_partitions(&four_partitions, partition);
+            let four_partitions = [
+                bo,
+                &BlockOffset{ x: bo.x + hbsw as usize, y: bo.y },
+                &BlockOffset{ x: bo.x, y: bo.y + hbsh as usize },
+                &BlockOffset{ x: bo.x + hbsw as usize, y: bo.y + hbsh as usize }
+            ];
+            let partitions = get_sub_partitions(&four_partitions, partition);
 
-        rd_cost += partitions.iter().map(|&offset| {
-                encode_partition_bottomup(
+            partitions.iter().for_each(|&offset| {
+                if let (cost, Some(mode_decision)) = encode_partition_bottomup(
```

Comment:
Is this called only when `partition == PartitionType::PARTITION_SPLIT`? I am not seeing the condition, so at first glance it seems like we test this four-way split many times. If this is needed could you please add a comment explaining this part?

Reply:
No, it is called for other partition types, like horz and vert as well.
Previously before this patch, i,e. w/o horz and vert,  encode_partition_bottomup() is given the four quadrants of current block. However, with hor and vert partition types enabled, it can fed with, two rectangles of current block, for ex 16x8 and 8x16 for the block 16x16, for the case PARTITION_HORZ. But note that in that case,  encode_partition_bottomup( ) will set can_split false because it is not square and skip the loop for each partition types. Also, if the rectangle straddles on the frame border, it should not have passed to encode_partition_bottomup( ) by the above checking of split_horz or split_vert.

Reply:
Right, it works as expected.